### DISCUSSION TITLE
Architecting InnateMC

### DISCUSSION THREAD

**Discussion by heftymouse** [InnateMC/InnateMC#25] [RFCs]
As this app becomes larger, it becomes important to define a standardized architecture so that it remains easy to maintain in the future. This is dependent on opinions and peoples' conventions so it'd be good to discuss it in, well, this discussion.

**Comment by heftymouse** [comment]
Here's what I had so far:
- It's unlikely that we ever move away from macOS or SwiftUI so it's unnecessary to excessively decouple functionality and implementation.
- SwiftUI has built-in support for two-way data binding so having a separate viewmodel is unnecessary.
- Body and event/completion handlers should be in separate parts of the view.
~~- Global 'services' should be passed in through a singleton EnvironmentObject~~
- 'Services' should be separate singleton classes.

**Reply by BoogieMonster1O1** [reply]
> Here's what I had so far:
> 
> * It's unlikely that we ever move away from macOS or SwiftUI so it's unnecessary to excessively decouple functionality and implementation.
> * SwiftUI has built-in support for two-way data binding so having a separate viewmodel is unnecessary.
> * Body and event/completion handlers should be in separate parts of the view.
> * Global 'services' should be passed in through a singleton EnvironmentObject

1 - Agreed. We don't need to overengineer that part.
2 - That can't be moved into `ContentView` because there will be multiple `ContentView` instances (multiple windows, tabs). That's why that's the global environment object
3 - I feel this is largely unnecessary. It might make the view look better but also might make it harder to understand; especially for larger views with lots of handlers.
4 - Elaborate on \`services`

**Reply by heftymouse** [reply]
2 - The current 'viewmodel' is an incorrectly named DI container. A viewmodel is an encapsulation of a view's state and commands and 'adapts' the model data so that it can be consumed by the view. The point was not made to suggest a change right now, but as a guideline to be followed in the future.

3 - The suggestion was made with the intent of making it easier to understand. Separating handlers allows the reader to more easily follow the flow of the UI and intuitively understand what it's doing. It also allows more than one person to work on separate parts of a view without constantly running into tricky merge conflicts. If having too many handlers is an issue perhaps it should be factored into separate views.

4 - This isn't a proper definition but here I'm calling objects that provide global functionality that can be single instanced services.

**Reply by BoogieMonster1O1** [reply]
2 - It's just shared data, we can rename it to that too. Otherwise, something along the lines of DI container is acceptable too
3 - Fair enough, we can do that for newer views and update existing ones on a rolling basis

**Reply by BoogieMonster1O1** [reply]
> 'Services' should be separate singleton classes.

Can you elaborate? Are you suggesting using the environment over an environment object?

**Reply by heftymouse** [reply]
Yes

**Reply by BoogieMonster1O1** [reply]
What would be the advantage over utilizing an Environment object with published fields?

**Reply by heftymouse** [reply]
It removes one object from the heap I suppose

**Reply by BoogieMonster1O1** [reply]
> It removes one object from the heap I suppose

At the cost of five more, the environment keys

**Reply by heftymouse** [reply]
The environment object will have those five as members anyways

**Reply by BoogieMonster1O1** [reply]
> The environment object will have those five as members anyways

Environment Object - 1 encapsulating class + 5 members
Environment - 5 environment keys + 5 members

Removing an object from the heap isn't a valid argument lol

**Comment by BoogieMonster1O1** [comment]
InnateKit is starting to sound like over-engineering. It has a couple of abstractions for things that we could remove. The advantages of InnateKit are as follows.
- Easier testing (counter - test driven development is a bad idea)
- Possibility to create a CUI launcher (counter - not on InnateMC's plans for the foreseeable future)

InnateKit and InnateMC are also starting to have some blurred lines, namely preferences, as it is not known in which module they should be stored. 

The merger of InnateKit and InnateMC would signal the end of a hybrid development architecture and the beginning of a monolithic architecture.

**Reply by heftymouse** [reply]
I agree, having them as separate targets is pointless given the direction we're going in.

**Reply by BoogieMonster1O1** [reply]
#27 addresses this
